src/main/java/de/metas/ui/web/material/cockpit/MaterialCockpitRowRepository.java

```diff
@@ -152,33 +181,50 @@ private boolean retrieveIsIncludePerPlantDetailRows()
 
 	private List<I_M_Product> retrieveRelevantProducts(@NonNull final List<DocumentFilter> filters)
 	{
-		final int orgId = Env.getAD_Org_ID(Env.getCtx());
-		final List<I_M_Product> allProducts = orgIdToproducts
-				.getOrLoad(orgId, () -> retrieveAllProducts(orgId));
+		final OrgId orgId = OrgId.ofRepoIdOrAny(Env.getAD_Org_ID(Env.getCtx()));
+
+		final int limit = Services
+				.get(ISysConfigBL.class)
+				.getIntValue(SYSCONFIG_EMPTY_PRODUCTS_LIMIT, DEFAULT_EMPTY_PRODUCTS_LIMIT);
 
-		return allProducts.stream()
-				.filter(materialCockpitFilters.toProductFilterPredicate(filters))
-				.collect(ImmutableList.toImmutableList());
+		final CacheKey cacheKey = new CacheKey(
+				orgId,
+				ProductFilterUtil.extractProductFilterVO(filters),
+				limit);
+
+		final List<I_M_Product> products = productFilterVOToProducts
+				.getOrLoad(cacheKey, () -> retrieveProductsFor(
+						cacheKey.getOrgId(),
+						cacheKey.getProductFilterVO(),
+						cacheKey.getLimit()));
+
+		return products;
 	}
 
-	private List<I_M_Product> retrieveAllProducts(final int adOrgId)
+	private List<I_M_Product> retrieveProductsFor(
```

Comment:
u could have only one param, the CacheKey

Reply:
i though about that too...yes, doing it..

Reply:
done ✔️